### PR TITLE
Add TensorFlow module and extensions

### PR DESCRIPTION
**Pull request by trevor-m**
This PR adds TensorFlow 2 support to Transformer Engine. It was developed by @kaixih @wenscarl and myself.

The extensions are built using CMake in a way that is similar to Jax and can be enabled using:
```
NVTE_FRAMEWORK=tensorflow pip install .
```

### CODE REVIEW

**Comment by ptrendx** [comment]
@trevor-m please sign your commits (see CONTRIBUTING.rst for details)

File: transformer_engine/tensorflow/csrc/extensions.cu [hunk 1]
```diff
@@ -0,0 +1,1140 @@
...
+
+  nvte_cast_transpose_dbias(input_cu.data(), cast_output_cu.data(),
+                            transposed_output_cu.data(), dbias_cu.data(),
+                            workspace.data(), stream);
+
+  // Fill workspace
+  auto w_s = workspace.shape();
+  std::vector<size_t> w_shape_vec{w_s.data, w_s.data + w_s.ndim};
+
+  void* workspace_ptr = AllocateSpace(w_shape_vec, workspace.dtype());
```
Comment (timmoon10) [inline transformer_engine/tensorflow/csrc/extensions.cu:347]: Does the TensorFlow GPU allocator do buffer caching like in PyTorch? I'm a little suspicious since the stream logic only affects how we call `cudaMemset`. If we're allocating a new GPU buffer every time we call this operation, that would add a lot of overhead. It's not critical for the first pass since it won't affect correctness, but we'll need to figure this out when we want performance.

There are a few other dispatch functions in this file where this is relevant.

Comment (kaixih) [reply transformer_engine/tensorflow/csrc/extensions.cu:347]: TF GPU allocator allocates a large chunk of GPU memory and suballocates it using a best-fit with coalescing (BFC) allocator, which is quite efficient on "creating" and "freeing" memory. We have already checked the profiling results and didn't notice significant overhead on this.

**Comment by kaixih** [comment]
@timmoon10 It seems the CI/CD tests will be triggered manually. Can we make it automatic?

File: transformer_engine/tensorflow/recipe.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
+"""This module provides predefined FP8 recipes."""
```
Comment (ksivaman) [inline transformer_engine/tensorflow/recipe.py:1]: Why is this file necessary? Looks like it's *almost* copied from `transformer_engine/common/recipe.py`?

**Comment by timmoon10** [comment]
> @timmoon10 It seems the CI/CD tests will be triggered manually. Can we make it automatic?

We have to do some security precautions to run on our CI machines, so it's not really possible to run everything automatically. Feel free to post or ping me or @ksivaman whenever you want to run. If you really want to run whenever you want, message me on Slack and I can give you some instructions.

File: transformer_engine/tensorflow/recipe.py (hunk 2 above)
Comment (kaixih) [reply transformer_engine/tensorflow/recipe.py]: Yes, you are right. Fixed by reusing the existing `recipe.py`. PTAL.

**Comment by ksivaman** [comment]
@kaixih Could you sign-off your commits, see [contributing guide](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst) for details.

**Comment by kaixih** [comment]
> @kaixih Could you sign-off your commits, see [contributing guide](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst) for details.

Oh, just noticed this comment. I have already added a couple of commits. How do I fix them to add the sign off?

File: README.rst [hunk 3]
```diff
@@ -108,6 +108,34 @@ JAX
         # Update FP8 metas
         other_variables = te.update_fp8_metas(other_grads)
 
+.. code-block:: python
```
Comment (timmoon10) [inline README.rst:111]:
```suggestion

TensorFlow
^^^^^^^^^^

.. code-block:: python
```

**Review by timmoon10 — APPROVED** [review]
I haven't gotten through all of the examples and tests yet, but I don't see any major problems. Consider this a tentative approval. I'll work on getting the CI infrastructure working with these changes.

File: qa/L0_lint/test.sh [hunk 4]
```diff
@@ -15,12 +15,12 @@ then
   cpplint --root transformer_engine/common/include --recursive transformer_engine/common/include
   echo "Checking C++ files"
   cpplint --recursive --exclude=transformer_engine/common/include transformer_engine
-  cpplint --recursive transformer_engine/pytorch
+  cpplint --recursive transformer_engine/pytorch transformer_engine/tensorflow
```
Comment (timmoon10) [inline qa/L0_lint/test.sh:18]: This test is performed in `qa/L0_tensorflow_lint/test.sh`:
```suggestion
  cpplint --recursive transformer_engine/pytorch
```

File: qa/L0_lint/test.sh [hunk 5]
```diff
@@ -15,12 +15,12 @@ then
...
 fi
 if [ -z "${CPP_ONLY}" ]
 then
   cp $TE_PATH/qa/L0_lint/pylintrc $TE_PATH
   cd $TE_PATH
   echo "Checking Python files"
-  pylint --recursive=y transformer_engine/common transformer_engine/pytorch
+  pylint --recursive=y transformer_engine/common transformer_engine/pytorch transformer_engine/tensorflow
```
Comment (timmoon10) [inline qa/L0_lint/test.sh:25]: This test is performed in `qa/L0_tensorflow_lint/test.sh`:
```suggestion
  pylint --recursive=y transformer_engine/common transformer_engine/pytorch

File: transformer_engine/tensorflow/softmax.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        return self.scaled_softmax(inp, scale)
+
+    def forward_tf_softmax(self, inp: tf.Tensor, mask: tf.Tensor) -> tf.Tensor:
+        """Framework softmax"""
+        if self.input_in_float16 and self.softmax_in_fp32:
+            inp = tf.cast(inp, tf.float32)
+
+        if self.scale is not None:
+            inp = inp * self.scale
+        mask_output = self.mask_func(inp, mask) if mask is not None else inp
```
[the remark this answers is not in the document]
Comment (kaixih) [reply transformer_engine/tensorflow/softmax.py:178]: Done.

File: README.rst (hunk 3 above)
Comment (kaixih) [reply README.rst:114]: Done.

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -31,3 +31,27 @@ jobs:
...
+    runs-on: ubuntu-latest
+    container:
+      image: nvcr.io/nvidia/tensorflow:23.02-tf2-py3
+      options: --user root
+    steps:
+      - name: 'Checkout'
+        uses: actions/checkout@v3
+      - name: 'Build'
+        run: |
+          apt-get update && apt-get install -y ninja-dev pybind11-dev
```
[the remark this answers is not in the document]
Comment (kaixih) [reply .github/workflows/build.yml]: Done.

File: qa/L0_lint/test.sh (hunk 4 above)
Comment (kaixih) [reply qa/L0_lint/test.sh]: Done.

File: qa/L0_lint/test.sh (hunk 5 above)
Comment (kaixih) [reply qa/L0_lint/test.sh]: Done.

**Review by timmoon10 — APPROVED** [review]
CI tests are green.
